FileSystem.Unix: Directory.Delete: remove per item syscall.

tmds opened the pull request:
By recursing using FileSystemEnumerable we know the file type and
can omit the stat calls made by DirectoryInfo.Exists.

For the top level path, we can omit the call also and handle
non-directories when rmdir errno is ENOTDIR.

For the recursive case we can avoid recursion when the top level path rmdir
succeeds immediately.

FileSystemEntry is updated so IsSymbolicLink remembers the file is symbolic link
and does not make a syscall for it.

tmds committed:
FileSystem.Unix: Directory.Delete: remove per item syscall.

By recursing using FileSystemEnumerable we know the file type and
can omit the stat calls made by DirectoryInfo.Exists.

For the top level path, we can omit the call also and handle
non-directories when rmdir errno is ENOTDIR.

For the recursive case we can avoid recursion when the top level path rmdir
succeeds immediately.

FileSystemEntry is updated so IsSymbolicLink remembers the file is symbolic link
and does not make a syscall for it.

tmds commented:
CI fails are not related: `RegexMatchTests.Match_ExcessPrefix(engine: SourceGenerated)` times out on Windows and Linux.

adamsitnik commented:
As soon as we merge https://github.com/dotnet/runtime/pull/60214 I am going to review this PR.

tmds commented:
I've rebased and removed the overlap with https://github.com/dotnet/runtime/pull/60214. This is up for review.

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -504,17 +500,38 @@ private static void RemoveDirectoryInternal(DirectoryInfo directory, bool recurs
                     case Interop.Error.ENOENT:
-                        if (!throwOnTopLevelDirectoryNotFound)
+                        if (!topLevel)
                         {
-                            return;
+                            return true;
+                        }
```
This one's confusing me, so I'll try to explain the logic. Please correct me if I got it wrong:

The [rmdir docs](https://man7.org/linux/man-pages/man2/rmdir.2.html) say:

```
ENOENT A directory component in pathname does not exist or is a dangling symbolic link.
```

So if we get ENOENT, it means the current entry is a symlink. And only if the current entry is the top level, we avoid deleting the contents of the link _target_ recursively. We achieve this by returning `true`, to assume the operation succedeed.

Was that correct?


But then, why do we throw (by going to default) if the current entry is not the top level?

tmds replied:
> ENOENT A directory component in pathname does not exist or is a dangling symbolic link.

You should read this as:

> ENOENT A directory component in pathname does not exist or _A directory component in pathname_ is a dangling symbolic link.

When the full pathname is a symbolic link, dangling or not, `rmdir` returns `ENOTDIR`.

tmds replied:
I've added a comment.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -504,17 +500,38 @@ private static void RemoveDirectoryInternal(DirectoryInfo directory, bool recurs
+                    case Interop.Error.ENOTDIR:
+                        if (topLevel)
+                        {
+                            if (!DirectoryExists(fullPath))
```
I don't understand why `DirectoryExists` is called for `ENOTDIR`. According to the [manual](https://man7.org/linux/man-pages/man2/rmdir.2.html)

rmdir fails with:
> * `ENOENT` A directory component in pathname does not exist or is a dangling symbolic link
> * `ENOTDIR` - pathname, or a component used as a directory in pathname, is not, in fact, a directory

I would expect that:
* on `ENOENT` we check if it's dangling symbolic link. If it is, we remove the link.
* on `ENOTDIR` we just call `DeleteFile`

tmds replied:
See https://github.com/dotnet/runtime/pull/59520#discussion_r752447729.

tmds replied:
I've added a comment.

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -433,69 +434,64 @@ public static void MoveDirectory(string sourceFullPath, string destFullPath)
-                return;
-            }
+                foreach ((string childPath, bool isDirectory) in
+                            new FileSystemEnumerable<(string, bool)>(
+                                fullPath,
+                                static (ref FileSystemEntry entry) =>
+                                {
+                                    // Don't report symlinks to directories as directories.
+                                    bool isRealDirectory = !entry.IsSymbolicLink && entry.IsDirectory;
 
-            if (recursive)
-            {
-                try
+                                    return (entry.ToFullPath(), isRealDirectory);
+                                },
+                                EnumerationOptions.Compatible))
```
Tiny nit: Could you please save the `FileSystemEnumerable` instance into a variable, outside of the try? It won't enumerate until you enter the `foreach`, and it's easier to read and debug:

Note: GitHub is not allowing me to add this as a suggestion 😠 (maybe it's a very complicated diff?).
```cs
            FileSystemEnumerable<(string, bool)> entries = new(
                fullPath,
                static (ref FileSystemEntry entry) =>
                {
                        // Don't report symlinks to directories as directories.
                        bool isRealDirectory = !entry.IsSymbolicLink && entry.IsDirectory;
                    return (entry.ToFullPath(), isRealDirectory);
                },
                EnumerationOptions.Compatible);

            try
            {
                foreach ((string childPath, bool isDirectory) in entries)
```

carlossanlop reviewed (commented):
Looks good so far. I left some comments for you to consider. Thanks for the change, @tmds.

Can you please run the enumeration benchmarks to ensure there's no perf regression?

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -433,69 +434,64 @@ public static void MoveDirectory(string sourceFullPath, string destFullPath)
+            if (!TryRemoveDirectory(fullPath, topLevel: true, throwWhenNotEmpty: !recursive))
             {
-                throw Interop.GetExceptionForIoErrno(Interop.Error.ENOENT.Info(), fullPath, isDirectory: true);
+                RemoveDirectoryRecursive(fullPath);
```
this is confusing:
* a method named `Try*` can throw an exception instead of returning `false`
* it seems that the `recursive` argument is ignored and we always try a recursive removal when non-recursive has failed. I need to read `RemoveDirectoryRecursive` to find out it's not the case.

tmds replied:
I agree. What do you think about renaming `TryRemoveDirectory` to `TryRemoveEmptyDirectory`, and `RemoveDirectoryRecursive` to `RemoveDirectoryCore`?

tmds replied:
I've renamed the methods and added some comments.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -433,69 +434,64 @@ public static void MoveDirectory(string sourceFullPath, string destFullPath)
-                                    firstException = exc;
-                                }
-                            }
+                            RemoveDirectoryRecursive(childPath);
```
Would it be possible, to perform a single recursive enumeration using a single `FileSystemEnumerable` instance and avoid the need of having to call the method recursively?

tmds replied:
We can't do that because we need to delete from the leaves to the top.

tmds committed:
PR feedback

tmds committed:
test: use Directory.CreateSymbolicLink

tmds commented:
@adamsitnik @carlossanlop I've renamed the methods and added some comments. I hope it is more clear now what the implementation does. I still need to add the test.

adamsitnik reviewed (approved):
@tmds thank you for another great contribution!

I've benchmarked this PR for removing a copy of a `dotnet/runtime` repository with all the build artifacts.

On average it's 8% faster and allocates 50% less memory:

|                   Method |    Job |    Mean | Ratio | Allocated |
|------------------------- |------- |--------:|------:|----------:|
| RecursiveDeleteDirectory | before | 2.550 s |  1.00 |     78 MB |
| RecursiveDeleteDirectory |  after | 2.356 s |  0.92 |     34 MB |

Just out of curiosity I've compared .NET to the the [rm](https://github.com/coreutils/coreutils/blob/master/src/rm.c) core utility, .NET is still 10-20% slower.  Most likely because `rm` is not using recursive code (it uses `fts`):

https://github.com/coreutils/coreutils/blob/3e2d64448300ea91be258af5b64ed3b97b865bd6/src/rm.c#L17-L19

But the [implementation](https://github.com/coreutils/coreutils/blob/master/src/remove.c) is non-trivial ;)

tmds committed:
Run test against different Delete APIs

tmds commented:
@carlossanlop I've updated the test so it also runs against the DirectoryInfo.Delete API.
The PR should now be good to merge.

adamsitnik commented:
@tmds some of the tests are failing now. Could you PTAL?

tmds commented:
@adamsitnik it seems that this fails on Windows. Is that expected?
```cs
            var target = GetTestFilePath();
            Directory.CreateDirectory(target);

            var linkParent = GetTestFilePath();
            Directory.CreateDirectory(linkParent);

            var linkPath = Path.Combine(linkParent, GetTestFileName());
            Directory.CreateSymbolicLink(linkPath, target);

            Assert.True(Directory.Exists(linkPath), "linkPath should exist");
```

adamsitnik commented:
> it seems that this fails on Windows. Is that expected?

It's not expected to me, but I am not a symlink expert. @carlossanlop @Jozkee is that expected?

@tmds Since you have not touched the Windows part, could you please just mark this test as Unix-specific for now so we can merge the PR?

tmds committed:
test: update Exists check for Windows

tmds committed:
Fix tests

tmds commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -483,17 +479,38 @@ private static void RemoveDirectoryInternal(DirectoryInfo directory, bool recurs
+                        {
+                            if (!DirectoryExists(fullPath))
+                            {
+                                throw Interop.GetExceptionForIoErrno(Interop.Error.ENOENT.Info(), fullPath, isDirectory: true);
```
@adamsitnik I would prefer to change this error code because I think the previous behavior was unintentional.
If you try to delete a socket file using `Directory.Delete`, the code used to throw `ENOENT`.
I think it is more appropriate to throw `ENOTDIR`. Do you agree?

tmds replied:
@adamsitnik can you share your opinion?

tmds commented:
> It's not expected to me, but I am not a symlink expert. @carlossanlop @Jozkee is that expected?

nvmd, I had unintentionally made changes to another similar test which caused it to fail.

The test should pass now on Linux and Windows.
